File: 프로그래머스/2/258711. 도넛과 막대 그래프/도넛과 막대 그래프.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> solution(vector<vector<int>> edges) {
    vector<int> answer(4, 0);
    vector<vector<int>> graph(1000001);
    vector<int> indegree(1000001, 0), outdegree(1000001, 0);
    int max_node = 0;

    for (const auto& edge : edges) {
        int from = edge[0], to = edge[1];
        max_node = max({max_node, from, to});
        graph[from].push_back(to);
        outdegree[from]++;
        indegree[to]++;
    }

    // Find the central node
    for (int i = 1; i <= max_node; i++) {
        if (outdegree[i] >= 2 && indegree[i] == 0) {
            answer[0] = i;
            break;
        }
    }

    // Analyze each subgraph
    for (int node : graph[answer[0]]) {
        int nodes = 1, edges = 0;
        queue<int> q;
        vector<bool> visited(max_node + 1, false);
        bool is_cycle = false;
        
        q.push(node);
        visited[node] = true;
        
        while (!q.empty()) {
            int current = q.front();
            q.pop();
            
            for (int next : graph[current]) {
                edges++;
                if (visited[next]) {
                    if (next == node) is_cycle = true;
                    continue;
                }
                visited[next] = true;
                nodes++;
                q.push(next);
            }
        }
        
        if (is_cycle) {
            if (nodes == edges) {
                answer[1]++;  // Donut
            } else {
                answer[3]++;  // Figure-8
            }
        } else {
            answer[2]++;  // Stick
        }
    }

    return answer;
}

File: 프로그래머스/2/258711. 도넛과 막대 그래프/도넛과 막대 그래프.js


/*

[중간 정점, 도넛, 막대, 8자] 를 반환해야 한다

도넛모양 특징
- node수 = edge 수

막대모양 특징
- node 수 = edge 수 + 1

8자모양 특징
- node 수 = edge 수 - 1

*/



function bfs(node, graph){

  var visited = new Array(graph.length).fill(false);
  var edges = 0;
  var nodes = 1;
  var queue = [];
  queue.push(node);
  
  while(queue.length > 0){
    var current = queue.shift();
    visited[current] = true;
    if(graph[current]){
      for(let n of (graph[current])){
        edges++;
        if(!visited[n]){
          nodes++;
          queue.push(n);
        }
      }
    }
  }

  return [edges, nodes];
}

function solution(edges) {
  var answer = [0,0,0,0];
  var graph = {};
  var outgoingEdges = {};
  var incomingEdges = {};
  for(let [from, to] of edges){

    if(!graph[from]) graph[from] = [];
    graph[from].push(to);

    outgoingEdges[from] = (outgoingEdges[from] || 0) + 1;
    incomingEdges[to] = (incomingEdges[to] || 0) + 1;
    
    if(!outgoingEdges[to]) outgoingEdges[to] = 0;
    if(!incomingEdges[from]) incomingEdges[from] = 0;

  }

  var centerNode = 0;
  var maxEdges = 0;

  for(let [node, edges] of Object.entries(outgoingEdges)){
    if(outgoingEdges[node] >= 2 && !incomingEdges[node]) {
      centerNode = parseInt(node);
      break;
  }}

  answer[0] = centerNode;
  for(let node of graph[centerNode]){
    let [edgeNum, nodeNum] = bfs(node, graph);
    if(edgeNum === nodeNum) answer[1]++;
    else if(edgeNum < nodeNum) answer[2]++;
    else answer[3]++;
  }

  return answer;
}


solution([[2, 3], [4, 3], [1, 1], [2, 1]]);